### PR TITLE
Use conditional types to make types significantly safer and more robust.

### PR DESCRIPTION
**Pull request by seansfkelley**
## `CustomCommands`

Due to the nature of `update.extend`, `update` itself cannot be effectively typed in such a way as to get type safety with the default commands while also allowing arbitrary custom commands. The reason for this is that (1) `extend` is side-effecting, so we can't capture a new type from the return value and (2) custom commands are entirely arbitrary, so in order to permit them we would have to have extremely permissive typings along the lines of `(any) => any`, which are so broad as to be almost useless.

In order to get around this, the types require a type parameter for custom commands. This allows normal usage to proceed unhindered and well-typed, but custom usage will have to be specially-notated. The type that represents this concept is `CustomCommands`.

The reason that `CustomCommands` is a separate, branded type rather than simply `extends object` is due to type inference. If `update`'s argument were typed as `type Spec<T, C extends object>`, the type parameter `C` would behave as a catch-all. Any fields you provide _anywhere_ in the nested spec would be lumped together under some crazy inferred type for `C`, thereby undermining type safety by effectively being `any`. Instead, `CustomCommands`' branding represents an un-inferrable, opt-in type that can only be explicitly specified by the user.

You never need to declare an object of type `CustomCommands<T>`, i.e., the brand doesn't exist at runtime. For typing purposes the branding is "unwrapped" and doesn't add extra nesting or the branded field itself to any type. In normal usage, it is invisible.

The typings file itself includes a brief example of how to use `CustomCommands`, reproduced here for convenience:

```ts
// Usage with custom commands is as follows:

interface MyCommands {
  $foo: string;
}

update<Foo, CustomCommands<MyCommands>>(..., { $foo: "bar" });

// It is suggested that if you use custom commands frequently, you wrap and re-export a
// properly-typed version of `update`:

function myUpdate<T>(object: T, spec: Spec<T, CustomCommands<MyCommands>>) {
  return update(object, spec);
}
```

## About this PR

This PR uses conditional types from Typescript 2.9 to inspect the arguments passed to `update` and provide very precise typings. It comes at the cost of demoting the ease of use of custom commands, as described above, but I consider that justifiable because as the explanation notes, types were barely checked (if at all) before due to too-permissive custom command support!

The improvements from this PR extend not just to type safety when calling `update`, but also to operations like renaming refactors! This means that in many cases if you rename a field on some type, calls to `update` that would modify the value of that field are also detected and refactored. This is a huge devex improvement from before, where those fields would be left dangling and pointing to nothing, causing subtle issues in the future.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -1,47 +1,49 @@
...
+  | ((v: T) => T)
+
+type ArraySpec<T> =
+  | BaseSpec<Array<T>>
+  | { $push: T }
+  | { $unshift: T }
+  // Technically splice is variadic, but we stop after the third shape because that usage is very rare and we can't do variadic types here.
+  | { $splice: Array<[number] | [number, number] | [number, number, T]> }
+  // TODO: Wanted this to work...
+  // | { [index: number]: Spec<T> }
```
Comment (seansfkelley) [inline index.d.ts]: This doesn't work because [type aliases can't refer to themselves except as properties](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases). I don't think there's a way around that.

File: index.d.ts [hunk 2]
```diff
@@ -1,47 +1,49 @@
...
+  // TODO: This workaround doesn't seem to work though...
+  // | { [K in keyof Array<T>]?: Spec<T> }
```
Comment (seansfkelley) [inline index.d.ts]: This one compiles, but as far as I can tell it doesn't actually do anything. Writing a type that should match this union member appears to ignore it as if it doesn't exist, I guess because `Array<T>` doesn't have any keys for `keyof` to take...?

File: index.d.ts [hunk 3]
```diff
@@ -1,47 +1,49 @@
...
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+declare function update<T>(target: T, spec: Spec<T>): T
+
+export type Spec<T> =
```
Comment (seansfkelley) [inline index.d.ts]: FYI, I renamed "query" -> "spec" to match the code.

File: index.d.ts [hunk 4]
```diff
@@ -1,47 +1,49 @@
...
+  T extends (Array<infer U> | ReadonlyArray<infer U>) ? ArraySpec<U> :
+  T extends (Map<infer K, infer V> | ReadonlyMap<infer K, infer V>) ? MapSpec<K, V> :
+  T extends (Set<infer U> | ReadonlySet<infer U>) ? SetSpec<U> :
+  T extends object ? ObjectSpec<T> :
+  BaseSpec<T>
+
+type BaseSpec<T> =
+  | { $set: T }
+  | { $apply: (v: T) => T }
+  | ((v: T) => T)
```
Comment (seansfkelley) [inline index.d.ts]: I tried putting something like the original custom command types right here:

```ts
| { [customCommand: string]: any }
```

But that basically turned off all type checking. That type matches pretty much everything, so if Typescript can't find a well-typed match, it'll pick this union member and then all bets are off. That's especially unfortunate since a huge part of the value of types here is to catch rename-refactor changing the shape of `T`, which this effectively disables.

I tried two methods to resolve this, but neither seemed to work. Perhaps I was turning the type system knobs incorrectly.

1.  Defining another type parameter the explicitly names the custom keys.

    ```ts
    declare function update<T, K extends PropertyKey = never>(target: T, spec: Spec<T, K>): T

    // Punching K around through `Spec` where necessary, then...

    type BaseSpec<T, K extends PropertyKey> =
      ...
      | { [customCommand: K]: any }
    ```

    Typescript didn't like this because `K` is not literally `string`, even if you made it so `K extends string`, so instead you can:

    ```ts
    declare function update<T, K extends string = never>(target: T, spec: Spec<T, K>): T

    type BaseSpec<T, K extends string> = // Has to be string
      ...
      | { [C in K]?: any }
    ```

    Which almost type checks, but if you punch it through to all the places (namely, `ObjectSpec`) where it needs to get passed, Typescript will start complaining that

    > Type alias 'Spec' circularly references itself.

    Which is surprising, since it can easily reference itself in `ObjectSpec<T>` with `{ [K in keyof T]?: Spec<T[K]> }`, but adding that extra `K` type parameter to `Spec` breaks it. Perhaps this is because that type parameter appears on the left side (as a key) rather than the right (as a value)?

2. Similar to above, but instead of using a type to express the keys, I used a type to express the whole set of well-typed operations that are legal.

    ```ts
    declare function update<T, C extends object = {}>(target: T, spec: Spec<T, C>): T

    type BaseSpec<T, C extends object> =
      ...
      | C
    ```

    This has the same problem as above. I actually prefer this method because it's more precise in allowing you to well-type your custom commands if you type-parameterize your `update` instance properly. I'm a little confused why this is illegal, since if you collapse out the noise you should get something like `Spec<T, C> = ... | C` which seems pretty legal. I'll have to dig into this one more.

An idea I haven't tried is to allow type-parameterizing the different types of operations independently (like `update<T, A extends ArraySpec<T> ...>`, etc.). This could get REAL tedious, but might work around the limitations of self-referentiality. There's a [Typescript PR in the works right now](https://github.com/Microsoft/TypeScript/pull/23696) that would make it more usable, too, by allowing you to just override the one you want (assuming it works at all).

As a design goal of these approaches, I was trying to get the general case (no customization) to work excellently out of the box, and then only if you want to customize it do you have to provide extra type information (seems reasonable, I think). That's why I really don't like the `{ [customCommand: string]: any }` method as an option.

Comment (seansfkelley) [reply index.d.ts]: Interestingly, it looks like this might be a heuristic around mutually recursive types failing. You can strip the failing example down to this:


```ts
export type Spec<T, C extends object> = ObjectSpec<T, C>

type ObjectSpec<T, C extends object> =
  | { [K in keyof T]?: Spec<T[K], C> }
  | C
```

But that `Spec` type is just a passthrough alias, so you can inline it because it isn't doing anything useful:

```ts
export type Spec<T, C extends object> = 
  | { [K in keyof T]?: Spec<T[K], C> }
  | C
```

At which point, it compiles successfully. So then of course I tried to inline the entire type into one gargantuan whole:

```ts
export type Spec<T, C extends object> = 
  | (
      T extends (Array<infer U> | ReadonlyArray<infer U>) ? (
        | { $push: T }
        | { $unshift: T }
        // Technically splice is variadic, but we stop after the third shape because that usage is very rare and we can't do variadic types here.
        | { $splice: Array<[number] | [number, number] | [number, number, T]> }
        // TODO: Wanted this to work...
        // | { [index: number]: Spec<T> }
        // TODO: This workaround doesn't seem to work though...
        // | { [K in keyof Array<T>]?: Spec<T> }
      ) :
      T extends (Map<infer K, infer V> | ReadonlyMap<infer K, infer V>) ? (
        | { $add: Array<[K, V]> }
        | { $remove: Array<K> }
      ) :
      T extends (Set<infer U> | ReadonlySet<infer U>) ? (
        | { $add: Array<T> }
        | { $remove: Array<T> }
      ) :
      T extends object ? (
        | { $toggle: Array<keyof T | number> }
        | { $unset: Array<keyof T | number> }
        | { $merge: Partial<T> }
        | { [K in keyof T]?: Spec<T[K], C> }
      ) : {}
    ) 
  | { $set: T }
  | { $apply: (v: T) => T }
  | ((v: T) => T)
  | C
```

But this doesn't compile, for the same reason. What gives?!

**Comment by andreiglingeanu** [comment]
> But that basically turned off all type checking. That type matches pretty much everything, so if Typescript can't find a well-typed match, it'll pick this union member and then all bets are off.

That was my conclusion when I started the work on writing strong types for the library, it was so especially then when conditional types weren't a thing yet. I think getting such types is quite hard because users have a very wide range of use-cases with lots of different data types

I'm not too much of a TypeScript writer these days so I won't be able to give constructive feedback, maybe @andy-ms or @sandersn could

Nonetheless this looks pretty solid!

**Comment by seansfkelley** [comment]
Thanks, at least now I know I'm not crazy. :)

I played around with it some more, and it looks like Typescript 2.9 allows type aliases to do the things we want them to do here, so I updated the PR and I think the issues are cleared up! I just want to write some tests (whether or not I commit them) to gain confidence and then decide if Typescript 2.9 is too new for consumers (I believe DefinitelyTyped waits one minor release cycle before allowing new features to be used in the typings).

File: index.d.ts [hunk 5]
```diff
@@ -1,47 +1,47 @@
...
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+declare function update<T, C extends object = never>(target: T, spec: Spec<T, C>): T
+
+export type Spec<T, C extends object> =
```
Comment (seansfkelley) [inline index.d.ts]: `Query` -> `Spec` to match the names in the implementation.

File: index.d.ts [hunk 6]
```diff
@@ -1,47 +1,47 @@
...
-  | ((old: T) => any)
-type ArrayOperators<T> =
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+declare function update<T, C extends object = never>(target: T, spec: Spec<T, C>): T
```
Comment (seansfkelley) [inline index.d.ts]: `C` is the custom commands object. You are required to explicitly type-parameterize this method in order to gain access to your commands.

I could move `C` as the first type argument to make such parameterization more ergonomic, but potentially more confusing at first, as having the object being operated on as the first parameter seems like a better ordering. [Named assignment type parameters](https://github.com/Microsoft/TypeScript/pull/23696) is also coming down the pipe, so this issue will be moot in 3.0 anyway.

This must default to `never` because `{}` is basically the "don't typecheck" value that matches anything you throw at it in this context.

**Comment by andreiglingeanu** [comment]
You're definitely not crazy, I was having the same feelings when digging this thing (:

For the automatic testing: we already integrated the `test-dtslint` npm script into the main CI flow and it does provide _some_ types linting. But it definitely does not offer an exhaustive coverage. That's a start, I believe...

As for using the bleeding edge TS features, yeah, you're probably right, this is a bit too early. And we have two options: we can just do like DefinitelyTyped does and just wait for `2.10.*` or we can just give up and prepare for a bunch of new issues (those would be resolved by just telling the users to update their TS compiler).

**Comment by seansfkelley** [comment]
It's probably best to wait until 3.0, which should be coming out this month. I guess we'll keep this PR on ice until then!

File: index.d.ts [hunk 7]
```diff
@@ -1,55 +1,86 @@
...
+//   function myUpdate<T>(object: T, spec: Spec<T, CustomCommands<MyCommands>>) {
+//     return update(object, spec);
+//   }
+declare function update<T, C extends CustomCommands<object> = never>(
+  target: T,
+  spec: Spec<T, C>,
+): T;
+
+// See the comment on `update` for an explanation of this type.
+export type CustomCommands<T> = T & { __noInferenceCustomCommandsBrand: any };
```
Comment (seansfkelley) [inline index.d.ts]: This is somewhat unfortunate but not a dealbreaker, I don't think. I'm not sure what the best place to include the lengthy explanatory comment is (Github comment? Code comment?).

**Comment by seansfkelley** [comment]
Ping @andreiglingeanu!

Typescript 3.0 has been out for a bit. I updated the PR with a whole bunch more tests and moved the explanatory docs to here, preferring to leave just a short usage comment and link back here for the justification of why.

File: index.d.ts (hunk 6 above)
Comment (andreiglingeanu) [reply index.d.ts]: Will the users have to always pass the second argument, even if they don't use custom commands?

**Comment by andreiglingeanu** [comment]
I like this a lot!

I love that users can pass type info about their custom built command, if they want to get type safety.

We'll need to make sure this doesn't break existing calls where no custom commands are being used. Primarily this is because there was a lot of issues about broken TS builds, after we published typings (which were not as good as yours).

Also, do the tests pass on your side? I think they do not on mine

**Comment by seansfkelley** [comment]
Thanks!

> We'll need to make sure this doesn't break existing calls where no custom commands are being used. 

Assuming that users are using it as intended, i.e., mapping from T -> T rather than T -> U, things should be fine. While the types were very permissive before, any compilation errors not related to custom commands _should_ be legitimate bugs on the consumers' part. ~~That said, I'm going to add a few more tricky tests.~~ I added some tests for complex nested objects and union/intersection types, as well as a couple edge cases which are erroneously too-permissive and documented them as such. Hopefully this covers all practical usages!

> Also, do the tests pass on your side? I think they do not on mine

There were a few commits in the middle where they were broken but as of this writing they pass (as they do on the build machines). What issues are you seeing?

**Comment by seansfkelley** [comment]
Ping @kolodny: what do I need to do to get this PR merged?

**Comment by kolodny** [comment]
I defer all typescript related things to @andreiglingeanu. If he's good with this then by all means, let's merge and push a release

**Comment by andreiglingeanu** [comment]
@seansfkelley I was looking at the wrong branch from your repo, sorry. Now everything is good on my side, I also see that all the tests pass and this gives more type safety than what we previously had.

Let's merge this.

Also, I wonder what will happen to TS <2.9 users, they will just not get type information about the lib?

**Comment by seansfkelley** [comment]
TS 2.8 might work somewhat; I'm not sure. It has all the necessary syntax, but seems to be worse at enforcing/inferring the necessary types.

Anything below 2.8 won't work because they don't have conditional types. I'm following the model of DefinitelyTyped, which is to not distribute changes with new language features until another minor version has passed. We're 2 minor versions beyond, so I figure it's fine.

As a backup plan, such users could always depend on the the version immediately prior to this one.

**Comment by kolodny** [comment]
Awesome! I'm ready to merge, I'd like to push out a release for this too. What should the appropriate release notes for this be? I can use `Use conditional types to make types significantly safer and more robust.` if that describes this change well enough

**Comment by seansfkelley** [comment]
I think that's a good summary, yeah. Just note that (1) it only refers to Typescript, not Flow and (2) it requires Typescript >= 2.9 (which as I stated above I think is reasonable; ping @andreiglingeanu if you have any particular objections to doing that).
